nova/virt/vmwareapi/vmops.py

```diff
@@ -2253,46 +2406,161 @@ def _relocate_vm(self, vm_ref, context, instance, network_info,
+            'cacert': CONF.vmware.ca_file,
+            'insecure': CONF.vmware.insecure,
+            'pool_size': CONF.vmware.connection_pool_size
+        })
+        return self._session_cls(**vmware_conf)
+
+    def _do_migrate_disk(self, context, vm_ref, instance, dest):
+        """Copies the VM to the destination by doing a clone.
+
+        If the destination is another vCenter service, it creates a new
+        VMwareAPISession to the destination vCenter service to gather the
+        needed info for cloning the vm: datastore, service locator, folder.
+        """
+        image_meta = instance.image_meta
+        storage_policy = self._get_storage_policy(instance.flavor)
+        allowed_ds_types = ds_util.get_allowed_datastore_types(
+            image_meta.properties.hw_disk_type)
+
+        # Determine whether we should connect to the dest vCenter endpoint
+        # (if the instance is to be migrated outside of current vCenter)
+        remote = self._is_remote(dest)
+
+        service_spec = None
+        if remote:
+            session = self._create_dest_session(dest)
+            # set the instance_uuid needed in _get_service_locator_spec
+            dest.update({
+                'instance_uuid':
+                    session.vim.service_content.about.instance_uuid})
+            service_spec = self._get_service_locator_spec(dest)
+            cluster = vutil.get_moref(dest['cluster'],
+                                      'ClusterComputeResource')
+            datastore_regex = dest['datastore_regex']
+            # For migration to other vCenter service the disk_move_type
+            # needs to be moveAllDiskBackingsAndDisallowSharing
+            disk_move_type = "moveAllDiskBackingsAndDisallowSharing"
+        else:
+            session = self._session
+            cluster = self._cluster
+            datastore_regex = self._datastore_regex
+            disk_move_type = "moveAllDiskBackingsAndAllowSharing"
+
+        res_pool = vm_util.get_res_pool_ref(session, cluster)
+        datastore = ds_util.get_datastore(session, cluster,
+                                          datastore_regex,
+                                          storage_policy,
+                                          allowed_ds_types)
+        dc_info = self.get_datacenter_ref_and_name(datastore.ref)
+        folder = self._get_project_folder(dc_info, instance.project_id,
+                                          'Instances', session=session)
+
+        rel_spec = vm_util.relocate_vm_spec(self._session.vim.client.factory,
+                                            disk_move_type=disk_move_type,
+                                            res_pool=res_pool, folder=folder,
+                                            datastore=datastore.ref,
+                                            service=service_spec)
+        cloned_vm = self._clone_vm(vm_ref, rel_spec, folder,
+                                   name=self._get_migration_vm_name(instance))
```

Comment:
At this point the original VM is powered off, has volume detached (in vmware side) and NICs removed. An exception, including a clone failure, will put the VM in an ERROR state (nova manager does that).
If we are to handle the clone exception here, what do we do with it? Are we going to rollback everything (volume attachments, insance power, and network interfaces), and then raise an `InstanceFaultRollBack` which will instruct nova to keep the VM state as ACTIVE? This sounds like an option, if we know that clone is likely to fail. Is there any way to predict a failing clone in the first place?

Reply:
We need to try and put the VM in a state, so that the user can resume operations. If it's in error state, they will try to reset the state and start it. Which will not have the effect the user wants, with removed NICs and volumes.

`InstanceFaultRollback` would be the best option. We should only raise it, if we could really revert all our changes. Otherwise, we should at least do our best.

Regarding preventing the clone failure, I'm all with you, but we only have a couple of cases we know of and tickets for that already. It can still always happen, that something on the target host prevents our cloning efforts.

Reply:
still open ...?

Reply:
There were some rollbacks introduced in `migrate_disk_and_power_off` which catches any exception in this `_do_migrate_disk_` method and rolls back.